File: apps/api-node/src/utils/api-response.ts
export function ok<T>(data: T, message?: string) {
  return {
    success: true,
    data,
    message
  };
}

export function created<T>(data: T, message?: string) {
  return {
    success: true,
    data,
    message
  };
}

export function fail(error: string, message?: string, detail?: unknown) {
  return {
    success: false,
    error,
    message,
    detail
  };
}

File: apps/java-admin-service/src/main/java/vn/edu/fpt/lnfs/dto/ApiResponse.java
package vn.edu.fpt.lnfs.dto;

public record ApiResponse<T>(boolean success, T data, String error, String message) {
  public static <T> ApiResponse<T> ok(T data) {
    return new ApiResponse<>(true, data, null, null);
  }

  public static <T> ApiResponse<T> ok(T data, String message) {
    return new ApiResponse<>(true, data, null, message);
  }

  public static ApiResponse<Void> error(String error, String message) {
    return new ApiResponse<>(false, null, error, message);
  }
}
